tests/framework/extensions/nodes/ec2/ec2.go

```diff
@@ -19,111 +19,232 @@ func CreateNodes(client *rancher.Client, numOfInstances int) ([]*nodes.Node, err
+							},
+						},
 					},
 				},
-			},
-		},
-	}
+			}
 
-	reservation, err := ec2Client.SVC.RunInstances(runInstancesInput)
-	if err != nil {
-		return nil, err
-	}
+			reservation, err := ec2Client.SVC.RunInstances(runInstancesInput)
+			if err != nil {
+				return nil, err
+			}
 
-	var listOfInstanceIds []*string
+			var listOfInstanceIds []*string
 
-	for _, instance := range reservation.Instances {
-		listOfInstanceIds = append(listOfInstanceIds, instance.InstanceId)
-	}
+			for _, instance := range reservation.Instances {
+				listOfInstanceIds = append(listOfInstanceIds, instance.InstanceId)
+			}
 
-	//wait until instance is running
-	err = ec2Client.SVC.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
-		InstanceIds: listOfInstanceIds,
-	})
-	if err != nil {
-		return nil, err
-	}
+			//wait until instance is running
+			err = ec2Client.SVC.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
+				InstanceIds: listOfInstanceIds,
+			})
+			if err != nil {
+				return nil, err
+			}
 
-	//wait until instance status is ok
-	err = ec2Client.SVC.WaitUntilInstanceStatusOk(&ec2.DescribeInstanceStatusInput{
-		InstanceIds: listOfInstanceIds,
-	})
-	if err != nil {
-		return nil, err
-	}
+			//wait until instance status is ok
+			err = ec2Client.SVC.WaitUntilInstanceStatusOk(&ec2.DescribeInstanceStatusInput{
+				InstanceIds: listOfInstanceIds,
+			})
+			if err != nil {
+				return nil, err
+			}
 
-	// describe instance to get attributes=
-	describe, err := ec2Client.SVC.DescribeInstances(&ec2.DescribeInstancesInput{
-		InstanceIds: listOfInstanceIds,
-	})
-	if err != nil {
-		return nil, err
+			// describe instance to get attributes=
+			describe, err := ec2Client.SVC.DescribeInstances(&ec2.DescribeInstancesInput{
+				InstanceIds: listOfInstanceIds,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			readyInstances := describe.Reservations[0].Instances
+
+			sshKey, err := nodes.GetSSHKey(config.AWSSSHKeyName)
+			if err != nil {
+				return nil, err
+			}
+
+			var ec2Nodes []*nodes.Node
+			for _, readyInstance := range readyInstances {
+				ec2Node := &nodes.Node{
+					NodeID:          *readyInstance.InstanceId,
+					PublicIPAddress: *readyInstance.PublicIpAddress,
+					SSHUser:         config.AWSUser,
+					SSHKey:          sshKey,
+				}
+				ec2Nodes = append(ec2Nodes, ec2Node)
+			}
+
+			client.Session.RegisterCleanupFunc(func() error {
+				return DeleteNodes(client, ec2Nodes)
+			})
+
+			return ec2Nodes, nil
+		}
 	}
 
-	readyInstances := describe.Reservations[0].Instances
+	return ec2Nodes, nil
+}
 
-	sshKey, err := nodes.GetSSHKey(ec2Client.Config.AWSSSHKeyName)
+func CreateWindowsNodes(client *rancher.Client, numOfInstancesWin int, testWindows bool) ([]*nodes.Node, error) {
+	ec2Client, err := client.GetEC2Client()
 	if err != nil {
 		return nil, err
 	}
-
 	var ec2Nodes []*nodes.Node
+	for _, config := range ec2Client.ClientConfig.AWSEC2Config {
+		if config.IsWindows && testWindows {
```

Comment:
If this function is to create a windows node, why do we need this if statement. I would like a clarification. Also why are you checking two bools? Seems redundant

Reply:
I have restructured the code so that this `if` statement is gone, and the check for `isWindows` makes more sense now. I was also able to strip out a TON of no longer needed code as a result of my refactor.